consensus/istanbul/backend/announce.go

```diff
@@ -17,405 +17,671 @@
-	HighFreqBeforeFirstPeerState AnnounceFrequencyState = iota
+// This function will request announce messages from a set of validator addresses from a peer
+func (sb *Backend) sendGetAnnounces(peer consensus.Peer, valAddresses []common.Address) error {
+	logger := sb.logger.New("func", "sendGetAnnounces")
+	valAddressesBytes, err := rlp.EncodeToBytes(valAddresses)
+	if err != nil {
+		logger.Error("Error encoding valAddresses", "valAddresses", common.ConvertToStringSlice(valAddresses), "err", err)
+		return err
+	}
 
-	// In this state, send out an announce message every 1 minute for the first 10 announce messages after the first peer is established.
-	// This is on the assumption that when this node first establishes a peer, the p2p network that this node is in may
-	// be partitioned with the broader p2p network. We want to give that p2p network some time to connect to the broader p2p network.
-	HighFreqAfterFirstPeerState
+	go peer.Send(istanbulGetAnnouncesMsg, valAddressesBytes)
+	return nil
+}
 
-	// In this state, send out an announce message every 10 minutes
-	LowFreqState
-)
+// This function will reply from a GetAnnounce message.  It will retrieve the requested announce messages
+// from it's announceMsgCache.
+func (sb *Backend) handleGetAnnouncesMsg(peer consensus.Peer, payload []byte) error {
+	logger := sb.logger.New("func", "handleGetAnnouncesMsg")
+	var valAddresses []common.Address
 
-const (
-	HighFreqTickerDuration = 1 * time.Minute
-	LowFreqTickerDuration  = 10 * time.Minute
-)
+	if err := rlp.DecodeBytes(payload, &valAddresses); err != nil {
+		logger.Error("Error in decoding valAddresses", "err", err)
+		return err
+	}
 
-// The sendAnnounce thread function
-func (sb *Backend) sendAnnounceMsgs() {
-	sb.announceWg.Add(1)
-	defer sb.announceWg.Done()
+	sb.cachedAnnounceMsgsMu.RLock()
+	defer sb.cachedAnnounceMsgsMu.RUnlock()
+	// TODO:  Add support for the AnnounceMsg to contain multiple announce messages within it's payload
+	for _, valAddress := range valAddresses {
+		if cachedAnnounceMsgEntry, ok := sb.cachedAnnounceMsgs[valAddress]; ok {
+			logger.Trace("Sending announce msg", "peer", peer, "valAddress", valAddress)
+			go peer.Send(istanbulAnnounceMsg, cachedAnnounceMsgEntry.MsgPayload)
+		}
+	}
 
-	// Send out an announce message when this thread starts
-	go sb.sendIstAnnounce()
+	return nil
+}
 
-	// Set the initial states
-	announceThreadState := HighFreqBeforeFirstPeerState
-	currentTickerDuration := HighFreqTickerDuration
-	ticker := time.NewTicker(currentTickerDuration)
-	numSentMsgsInHighFreqAfterFirstPeerState := 0
+// This function will generate the lastest announce msg from this node (as opposed to retrieving from the announceMsgCache) and then broadcast it to it's peers,
+// which should then gossip the announce msg message throughout the p2p network, since this announce msg's timestamp should be the latest among all of this
+// validator's previous announce msgs.
+func (sb *Backend) gossipAnnounce() error {
```

Comment:
It might be worth renaming this to `generateAndGossipAnnounce` to make it clear that a fresh announce will be generated and cached.